src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java

```diff
@@ -0,0 +1,127 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.randomized;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import se.liu.ida.hefquin.engine.queryplan.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizer;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.CostModel;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RewritingRule;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RuleApplication;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CostEstimationUtils;
+import se.liu.ida.hefquin.engine.utils.Pair;
+
+public class IterativeImprovementBasedQueryOptimizer implements QueryOptimizer {
+	
+	protected final QueryOptimizationContext context;
+	protected final StoppingConditionForIterativeImprovement condition;
+
+	
+	public IterativeImprovementBasedQueryOptimizer (final QueryOptimizationContext ctxt, final StoppingConditionForIterativeImprovement x) {
+		assert ctxt != null;
+		assert x != null;
+		context = ctxt;
+		condition = x;
+	}
+	
+	
+	@Override
+	public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimizationException {
+		return optimize( context.getLogicalToPhysicalPlanConverter().convert(initialPlan,false) );
+	}
+
+	public PhysicalPlan optimize( final PhysicalPlan initialPlan ) throws QueryOptimizationException {
+		// The best plan we have found so far. As we have only found one plan, it is the best one so far.
+		PhysicalPlan bestPlan = initialPlan;
+		PhysicalPlan currentPlan = bestPlan; // This variable will hold the plan which is currently being worked on.
+		
+		// Creating the completable future.
+		CostModel costModel = context.getCostModel();
+		CompletableFuture<Double> futureCost = costModel.initiateCostEstimation(bestPlan);
+
+		// Declaring other variables here, possibly giving the future some time to be realized with minimal waiting.
+		RewritingRule rwRule = null;
+		List<PhysicalPlan> neighbours;
+		
+		Double[] neighbourCosts;
+		List<Pair<PhysicalPlan,Double>> betterPlans = new ArrayList<>();
+		Pair<PhysicalPlan,Double> betterPlan = null;
+		Random rng = new Random();

```

Comment:
Rather than creating this variable as a local variable within this function, you can make it a member variable. Additionally, you can mark it as `final`.